Feature/reaching definitions

lacinoire opened the pull request.

Commit:
intellij and structure set-up

Commit:
setup work for true-liveness and register allocation

denis631 committed:
true liveness analysis for assignments and guards. need to fix the loops endless recursion

denis631 committed:
fix the conditional/guard problem for the liveness union rather than reset

denis631 committed:
true liveness: fixpoint endless recursion fix

denis631 committed:
register allocation done. tested comparing to the homework tasks

lacinoire committed:
reaching definitions

lacinoire committed:
fix reaching definitions + extend main analysis runner

lacinoire committed:
keep idea folder in repo

lacinoire committed:
cleanup print statements

lacinoire committed:
gitignore update

denis631 commented on analysis/analysis/ReachingDefinitionsAnalysis.java:
```diff
@@ -0,0 +1,101 @@
+        if (assignment.getLhs() instanceof Variable) {
+            Variable variable = (Variable) assignment.getLhs();
+            reachingDefinitions.removeIf((rd) ->
+                rd.a.toString().equals(variable.toString())
```
is it better to compare by ids?

lacinoire replied:
👍🏼

denis631 commented on analysis/analysis/ReachingDefinitionsAnalysis.java:
```diff
@@ -0,0 +1,101 @@
+            reachingDefinitions
+                .add(new Tupel<Variable, Long>(variable, assignment.getDest().getId()));
+        }
+        return reachingDefinitions;
```
not sure, but shouldn't you return always a new instance of the reachingDefinitions object?
Is this reachingDefinitions gonna be passed to other state or does it start with NULL?

lacinoire replied:
Ahh Java 😆
Thanks for spotting, this actually fixed my only remaining bug 🙂

Flakebi committed:
Fix some bugs

- Add .gitignore file (not complete)
- Add hashCode for expressions
- Fix SyntheticComputationDFS

denis631 commented on analysis/analysis/ReachingDefinitionsAnalysis.java:
```diff
@@ -0,0 +1,101 @@
+//        System.out.println(reachingDefinitions);
+//        System.out.println(lessoreq(oldRD, reachingDefinitions));
+//        System.out.println(lub(reachingDefinitions, oldRD));
+        if (!lessoreq(oldRD, reachingDefinitions)) {
```
don't you have it the other way around?
In the example it is written this way:
```
if (!lessoreq(newflow,oldflow)){
```
but you compare whether old is > than reachingDef. Maybe switching the places gonna fix your bug

lacinoire replied:
Your right, I used this the wrong way around! Because the Method was not actually doing less or equal 😅 Fixed it so the method name corresponds to what it does 🙂

lacinoire committed:
address pr comments

lacinoire committed:
fix lessoreq comparison

lacinoire committed:
keep compiler main function and more complex example file

denis631 committed:
- refactoring. Using lessOrEq and lub methods, for more consistency
- added return ret to the first file, otherwise, since no value is returned, there is "no need" to compute anything
